### ISSUE TITLE
How to prevent the neural training to stop too early ?

### ISSUE THREAD

**Issue by EtienneAb3d** [ModernMT/MMT#347]
As discussed in a past exchange, I encounter a problem with the stop criterion. 

I have 16M pairs of sentences (mainly usual MultiUN, Europarl, ..).  Usually, this takes about 5 to 7 days to train with the standard MMT neural configuration.

I made a chunk extraction on them (I can explain afterwards what I'm expecting), to build about 60M pairs of chunks (of course, with a lot of redundancies, and a lot of noise). I gave both sentences pairs and chunks pairs to the neural training. This makes about 69M entries.

Since the chunks are a bit more easy to learn, with the standard training, the learning rate was going down very quickly, and it stopped in only 2 days, having done only 1.4 epochs.

To force it to go slower, I tried this options:
1) to get at least 1 epoch before decreasing the learning rate (1M*64mini-batch=64M pairs):
--learning-rate-decay-start-at 1000000 
2) to force the learning rate to decrease slowly, evaluated only every 1/3 epoch
--learning-rate-decay-steps 300000

But, the stop criterion is still evaluated, whatever the learning rate, and it stopped even faster after only 24H, and only 0.74 epochs.

I would enjoy to have an option to get a stop criterion like "if learning rate less than a given value".

PS : I'm now trying with the option --validation-steps 300000, hope this will be sufficient.

**Comment by davidecaroselli** [comment]
Hi @EtienneAb3d 

I understand that you are trying to experiment a bit with MMT. I suggest to comment these two lines here: https://github.com/ModernMT/MMT/blob/master/src/decoder-neural/src/main/python/nmmt/NMTEngineTrainer.py#L376

Then just run `mvn install` from `src` folder and the termination policy will be disabled.

Let me know if this solves your issue,
Davide

**Comment by EtienneAb3d** [comment]
I will have a look at your code right today... Because, even with the latest options, it is still stopping after only 0.9 epoch. :-(

Can you explain the effect of `--validation-steps 300000` ? It seems is is always evaluating it every 10000.

> 2018-02-14 19:22:55,213 [INFO] - Validation at step 860000: loss = 1.68874, perplexity = 5.41267, accuracy = 67.9282
> 2018-02-14 19:37:00,486 [INFO] - Validation at step 870000: loss = 1.69482, perplexity = 5.44568, accuracy = 67.6885
> 2018-02-14 19:52:03,249 [INFO] - Validation at step 880000: loss = 1.67968, perplexity = 5.36384, accuracy = 67.8441
> 2018-02-14 20:06:44,865 [INFO] - Validation at step 890000: loss = 1.69314, perplexity = 5.43655, accuracy = 67.575
> 2018-02-14 20:21:11,904 [INFO] - Validation at step 900000: loss = 1.68227, perplexity = 5.37774, accuracy = 67.966
> 2018-02-14 20:21:11,904 [INFO] - Validation perplexity at step 900000 (epoch 0.84): 5.377740; new best: 5.377740
> 2018-02-14 20:36:05,635 [INFO] - Validation at step 910000: loss = 1.67904, perplexity = 5.36043, accuracy = 67.7516
> 2018-02-14 20:49:58,467 [INFO] - Validation at step 920000: loss = 1.67827, perplexity = 5.35626, accuracy = 67.9156
> 2018-02-14 21:05:03,789 [INFO] - Validation at step 930000: loss = 1.68239, perplexity = 5.37841, accuracy = 67.7978
> 2018-02-14 21:19:46,890 [INFO] - Validation at step 940000: loss = 1.67415, perplexity = 5.33425, accuracy = 68.0417
> 2018-02-14 21:33:57,772 [INFO] - Validation at step 950000: loss = 1.66993, perplexity = 5.3118, accuracy = 68.1006
> 2018-02-14 21:48:47,278 [INFO] - Validation at step 960000: loss = 1.67663, perplexity = 5.34751, accuracy = 67.8231
> 2018-02-14 22:04:01,416 [INFO] - Validation at step 970000: loss = 1.73391, perplexity = 5.66274, accuracy = 67.5161
>

**Comment by nicolabertoldi** [comment]
Validation is performed every "--validation-steps", but also every "--checkpoint-steps"

So if you actually wants to validate every 300K steps you have to modify that parameter as well

**Comment by nicolabertoldi** [comment]
could you say whether between steps 960000 and 970000 e re-start of the training has been done>
Maybe a modification of the batch_size occurred?

could you post the portion of the full log in this interval?

**Comment by EtienneAb3d** [comment]
I did nothing special. But, since the data are a mix of 16M sentences, and 60M extracted chunks, they are quite heterogeneous. It's not surprising that the training is a bit random, especially  with lr=1 at this time.
Here is the full log.
[training.log.zip](https://github.com/ModernMT/MMT/files/1726742/training.log.zip)

**Comment by nicolabertoldi** [comment]
If you want to be sure that at least one full epoch is performed
you could have a setting like this

--validation-steps 300000
--checkpoint-steps 300000
--lr_decay_start_at 300000
--n_avg_checkpoints 20    (this is the default)

According to your figures you have 68844969 sentences, i.e. 1075703 steps (with batch_size=64)
you will perform ~3.3 validations (=1075703/300000) per epoch; hence your training will continue at least until 6 epochs (=20/3.3). But probably the training will continue much more, because of the reduction of the learning rate which can decrement from step 300000.

By the way we have trained good engines with huge data without completing an epoch.
We think that the total amount of processed text is important rather that this data are seen just once or more times.
The variability of the data is also crucial.

**Comment by EtienneAb3d** [comment]
Some specific data, in low quantity, are mixed with more generic ones. To get a good model, on these specific data, I really need that they are process many times. More than this, I don't think that a model trained only with lr=1 would be of a good quality. And, in my very first attempt, even if the lr was finally low, it was decreased very very quickly, thus, again, certainly not a good model.

I will now try with this change in the code pointed above:
```
                        if self.optimizer.lr < 0.01 and not perplexity_improves:
                            break
```
And these options:
```
--learning-rate-decay-start-at 1000000 
--learning-rate-decay-steps 300000
--validation-steps 300000
--checkpoint-steps 300000
```

If you see some things to tune..

**Comment by davidecaroselli** [comment]
It seems good to me. As a general note, please keep in mind that the actual validation and checkpoint steps are choosen to be the min between the specified value and the steps in one epoch.

This is irrilevant for medium/large training sets, while in your case you may hit the limit if you want to push those values even further.

**Comment by EtienneAb3d** [comment]
@nicolabertoldi, I'm thinking to this sentence:
> By the way we have trained good engines with huge data without completing an epoch.

I do not have huge data in this experimentation, only 16M sentences. The 60M chunks pairs are extracted from these 16M data. They are not new sentences.

Here is an example from Europarl..

**FR sentence:** 
"je déclare reprise la session du Parlement européen qui avait été interrompue le vendredi 17 décembre dernier et je vous renouvelle tous mes vux en espérant que vous avez passé de bonnes vacances."
**EN sentence:** 
"I declare resumed the session of the European Parliament adjourned on Friday 17 December 1999, and I would like once again to wish you a happy new year in the hope that you enjoyed a pleasant festive period."

**Chunking:**
je déclare reprise
la session du Parlement européen
qui
avait été interrompue
le vendredi 17 décembre dernier
et
je
vous
renouvelle
tous mes vux
en
espérant
que
vous avez passé de bonnes vacances
.
I declare resumed
the session of the European Parliament adjourned on
Friday 17 December 1999
,
and
I would like once again
to wish you
a happy new year
in
the hope
that
you enjoyed
a pleasant festive period
.

**After analyze, these chunks pairs are kept (in the 60M):**
KEPT: je déclare reprise	/	I declare resumed
KEPT: le vendredi 17 décembre dernier	/	Friday 17 December 1999
KEPT: la session du Parlement européen	/	the session of the European Parliament adjourned on

**Comment by davidecaroselli** [comment]
Hi @EtienneAb3d 

I'm closing this issue because I think the main issue has been solved. On the other hand I think you will prefer publish any update on the other issue that is more in focus.

If you need any other suggestions on this please feel free to re-open it!

Cheers,
Davide
